allowed_types are not considered

if allowed_types are defined, only selectable items should be displayed!

```
self.request.form['portal_type'] = []
```
imho ^ removes the portal_type query parameter and disables allowed_types

removing that line solves the problem in "real life" but it seems this feature is not taken into account within the tests 

```
Failure in test test_popup_items (archetypes.referencebrowserwidget.tests.test_product.IntegrationTestCase)
Traceback (most recent call last):
  File "/Users/peter/workspace/python/parts/opt/lib/python2.7/unittest/case.py", line 327, in run
    testMethod()
  File "/Users/peter/workspace/buildout.coredev/src/archetypes.referencebrowserwidget/src/archetypes/referencebrowserwidget/tests/test_product.py", line 665, in test_popup_items
    self.assertEqual(len(ROWS.findall(body)), wanted_rows)
  File "/Users/peter/workspace/python/parts/opt/lib/python2.7/unittest/case.py", line 509, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Users/peter/workspace/python/parts/opt/lib/python2.7/unittest/case.py", line 502, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 3 != 7
```

https://github.com/plone/archetypes.referencebrowserwidget/commit/48af4e89cadf0b23357adacd0f212385615eb926?diff=unified

--- a/CONTRIBUTORS.txt
+++ b/CONTRIBUTORS.txt
@@ -7,3 +7,4 @@
  - Thomas Desvenan, thomasdesvenain@gmail.com - fixes and minor features
  - Peter Holzer, hpeter@agitator.com - query update/fix
  - Gauthier Bastien, gauthier@imio.be - bug fixes
+ - Harald Friessnegger, harald@webmeisterei.com - bug fixes

--- a/docs/CHANGES.txt
+++ b/docs/CHANGES.txt
@@ -13,7 +13,11 @@ Changelog
 
 - #16 fix browsing for widgets in case field has set ``allowed_types``
 
->>>>>>> ed36daf... undo changes introduced with a537138d8fbfdbacd571886dd3728debe45ea9ee
+  (Warning: changed return value of view/getResult from 
+  `[`brain1, brain2,..]``  to ``[dict(item=brain1,browsable=True,
+  referenceable=False),...]`` in case you customized popup.pt) 
+  [fRiSi]
+
 
 2.5.2 (2014-09-11)
 ------------------

--- a/docs/CHANGES.txt
+++ b/docs/CHANGES.txt
@@ -6,9 +6,14 @@ Changelog
 
 - Replaced ``jq`` usage with ``$``
   [keul]
+
 - Fixing reference sorting issue due to global function refactoring done
   on recent versions
   [keul]
+
+- #16 fix browsing for widgets in case field has set ``allowed_types``
+
+>>>>>>> ed36daf... undo changes introduced with a537138d8fbfdbacd571886dd3728debe45ea9ee
 
 2.5.2 (2014-09-11)
 ------------------

--- a/src/archetypes/referencebrowserwidget/browser/popup.pt
+++ b/src/archetypes/referencebrowserwidget/browser/popup.pt
@@ -180,7 +180,8 @@
             referenceable python:view.isReferencable(item);
             browsable python:allow_browse and view.isBrowsable(item);
             title_or_id python:view.title_or_id(item);
-            color view/discreetColor;">
+            color view/discreetColor;"
+        tal:condition="python: referenceable or browsable or not isNotSelf">
 
           <tr tal:define="oddrow repeat/item/odd"
           tal:attributes="class python:oddrow and 'even' or 'odd'">

--- a/src/archetypes/referencebrowserwidget/browser/popup.pt
+++ b/src/archetypes/referencebrowserwidget/browser/popup.pt
@@ -172,18 +172,18 @@
     <tbody>
       <tal:results tal:define="plone_view context/@@plone;
                                normalizeString nocall:plone_view/normalizeString;"
-                   tal:repeat="item batch">
+                   tal:repeat="info batch">
         <tal:row
         tal:define="
+            item info/item;
             uid python:view.getUid(item);
             isNotSelf python:view.isNotSelf(item);
-            referenceable python:view.isReferencable(item);
-            browsable python:allow_browse and view.isBrowsable(item);
+            referenceable info/referenceable;
+            browsable python:allow_browse and info['browsable'];
             title_or_id python:view.title_or_id(item);
-            color view/discreetColor;"
-        tal:condition="python: referenceable or browsable or not isNotSelf">
+            color view/discreetColor;">
 
-          <tr tal:define="oddrow repeat/item/odd"
+          <tr tal:define="oddrow repeat/info/odd"
           tal:attributes="class python:oddrow and 'even' or 'odd'">
 
         <td>
@@ -220,10 +220,10 @@
                   tal:content="title_or_id" />
           </tal:foldercheck>
           <a class="discreet"
-		     tal:condition="referenceable"
-			 onclick="window.open(jQuery(this).attr('href'));return false;"
-			 tal:attributes="href python:view.preview_url(item)">
-          	(<span i18n:translate="" i18n:domain="plone">View</span>)
+             tal:condition="referenceable"
+             onclick="window.open(jQuery(this).attr('href'));return false;"
+             tal:attributes="href python:view.preview_url(item)">
+            (<span i18n:translate="" i18n:domain="plone">View</span>)
           </a>
 
           <div tal:condition="widget/show_review_state">

--- a/src/archetypes/referencebrowserwidget/browser/view.py
+++ b/src/archetypes/referencebrowserwidget/browser/view.py
@@ -289,12 +289,31 @@ class ReferenceBrowserPopup(BrowserView):
                 self.request.form['sort_on'] = 'getObjPositionInParent'
 
             result = qc(search_catalog=self.widget.search_catalog)
+
         else:
             result = []
+
         b_size = int(self.request.get('b_size', 20))
         b_start = int(self.request.get('b_start', 0))
-        
-        return Batch(result, b_size, b_start, orphan=1)
+        return Batch(self._prepareResults(result), b_size, b_start, orphan=1)
+
+    def _prepareResults(self, result):
+        items_with_info = []
+        for item in result:
+            browse = self.isBrowsable(item)
+            ref = self.isReferencable(item)
+            if self.allowed_types:
+                # we only show allowed_types and objects needed for browsing
+                if not (ref or browse or not self.isNotSelf(item)):
+                    continue
+
+            items_with_info.append(dict(
+                item=item,
+                browsable=browse,
+                referenceable=ref,
+                ))
+
+        return items_with_info
 
     def breadcrumbs(self, startup_directory=None):
         assert self._updated

--- a/src/archetypes/referencebrowserwidget/browser/view.py
+++ b/src/archetypes/referencebrowserwidget/browser/view.py
@@ -18,7 +18,7 @@ from Products.Five import BrowserView
 try:
     # Zope >= 2.13
     from AccessControl.security import checkPermission
-    checkPermission   # pyflakes
+    checkPermission  # pyflakes
 except ImportError:
     from Products.Five.security import checkPermission
 
@@ -282,6 +282,7 @@ class ReferenceBrowserPopup(BrowserView):
             self.request.form['path'] = {
                 'query': '/'.join(folder.getPhysicalPath()),
                 'depth': 1}
+            self.request.form['portal_type'] = []
             if 'sort_on' in self.widget.base_query:
                 self.request.form['sort_on'] = self.widget.base_query['sort_on']
             else:
@@ -292,7 +293,7 @@ class ReferenceBrowserPopup(BrowserView):
             result = []
         b_size = int(self.request.get('b_size', 20))
         b_start = int(self.request.get('b_start', 0))
-
+        
         return Batch(result, b_size, b_start, orphan=1)
 
     def breadcrumbs(self, startup_directory=None):
@@ -373,5 +374,5 @@ class ReferenceBrowserPopup(BrowserView):
         site_properties = portal_properties.site_properties
         types_use_view = site_properties.typesUseViewActionInListings
         if item.portal_type in types_use_view:
-            return item.getURL()+'/view'
+            return item.getURL() + '/view'
         return item.getURL()

--- a/src/archetypes/referencebrowserwidget/tests/test_product.py
+++ b/src/archetypes/referencebrowserwidget/tests/test_product.py
@@ -12,7 +12,7 @@ from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile
 from Products.Five import BrowserView
 try:
     from Testing.testbrowser import Browser  # Zope >= 2.13
-    Browser   # pyflakes
+    Browser  # pyflakes
 except ImportError:
     from Products.Five.testbrowser import Browser  # Zope < 2.13
 
@@ -25,7 +25,7 @@ from Products.PloneTestCase.PloneTestCase import default_password
 from Products.PloneTestCase.PloneTestCase import portal_owner
 try:
     import plone.uuid
-    plone.uuid   # pyflakes
+    plone.uuid  # pyflakes
     import pkg_resources
     uuid_version = pkg_resources.get_distribution("plone.uuid").version
     if uuid_version < '1.0.2':
@@ -199,7 +199,7 @@ class PopupTestCase(PopupBaseTestCase):
         alternate_template = named_template_adapter(
             ViewPageTemplateFile('sample.pt'))
         zope.component.provideAdapter(alternate_template,
-                                      adapts=(BrowserView, ),
+                                      adapts=(BrowserView,),
                                       provides=INamedTemplate,
                                       name='alternate')
         fieldname = 'multiRef2'
@@ -319,7 +319,7 @@ class PopupTestCase(PopupBaseTestCase):
         # now testing what URL is get for content's where "/view" if forced
         site_properties = self.portal.portal_properties.site_properties
         site_properties.typesUseViewActionInListings = ('RefBrowserDemo',)
-        assert popup.preview_url(brain) == brain.getURL()+'/view'
+        assert popup.preview_url(brain) == brain.getURL() + '/view'
 
     def test_at_url(self):
         makeContent(self.folder, portal_type='RefBrowserDemo', id='with space')
@@ -654,13 +654,15 @@ class IntegrationTestCase(FunctionalTestCase):
         assert '<input type="hidden" name="at_url" value="plone/demo1" />' in body
 
     def test_popup_items(self):
+        wanted_rows = 7
         wanted_insertlinks = 2
 
         body = self.getNormalizedPopup()
         INSERTLINK = re.compile(r'<input type="checkbox" class="insertreference" id="[0-9a-f]*?" rel="[0-9a-f]*?" />')
         INSERTLINK_UUID = re.compile(r'<input type="checkbox" class="insertreference" id="[\w]{8}-[\w]{4}-[\w]{4}-[\w]{4}-[\w]{12}" rel="[\w]{8}-[\w]{4}-[\w]{4}-[\w]{4}-[\w]{12}" />')
 
-        ROWS = re.compile(r'<tr.*?>(.*?)</tr>', re.MULTILINE|re.DOTALL)
+        ROWS = re.compile(r'<tr.*?>(.*?)</tr>', re.MULTILINE | re.DOTALL)
+        self.assertEqual(len(ROWS.findall(body)), wanted_rows)
         if HAS_DASH_UUID:
             self.assertEqual(len(INSERTLINK_UUID.findall(body)), wanted_insertlinks)
         else:
@@ -669,6 +671,7 @@ class IntegrationTestCase(FunctionalTestCase):
         makeContent(self.portal, portal_type='News Item', id='newsitem')
         body = self.getNormalizedPopup()
 
+        self.assertEqual(len(ROWS.findall(body)), wanted_rows + 1)
         if HAS_DASH_UUID:
             self.assertEqual(len(INSERTLINK_UUID.findall(body)), wanted_insertlinks)
         else:

--- a/src/archetypes/referencebrowserwidget/tests/test_product.py
+++ b/src/archetypes/referencebrowserwidget/tests/test_product.py
@@ -251,7 +251,7 @@ class PopupTestCase(PopupBaseTestCase):
         batch = popup.getResult()
         assert isinstance(batch, Batch)
         assert len(batch) == 1
-        assert batch[0].getObject() == self.portal.news
+        assert batch[0]['item'].getObject() == self.portal.news
         assert popup.has_queryresults
 
     def test_path_query(self):
@@ -267,8 +267,8 @@ class PopupTestCase(PopupBaseTestCase):
         assert isinstance(batch, Batch)
         # expected to have both the folder at "path" and its contents
         assert len(batch) == 2
-        assert batch[0].getObject() == self.portal.events
-        assert batch[1].getObject() == self.portal.events.aggregator
+        assert batch[0]['item'].getObject() == self.portal.events
+        assert batch[1]['item'].getObject() == self.portal.events.aggregator
         assert popup.has_queryresults
 
     def test_noquery(self):
@@ -676,7 +676,7 @@ class IntegrationTestCase(FunctionalTestCase):
             self.assertEqual(len(INSERTLINK_UUID.findall(body)), wanted_insertlinks)
         else:
             self.assertEqual(len(INSERTLINK.findall(body)), wanted_insertlinks)
-            
+
         # add a document, this will be addable in the popup
         makeContent(self.portal, portal_type='Document', id='another-doc')
         body = self.getNormalizedPopup()
@@ -685,7 +685,7 @@ class IntegrationTestCase(FunctionalTestCase):
             self.assertEqual(len(INSERTLINK_UUID.findall(body)), wanted_insertlinks + 1)
         else:
             self.assertEqual(len(INSERTLINK.findall(body)), wanted_insertlinks + 1)
-            
+
     def test_bc_navigationroot(self):
         makeContent(self.portal.folder1, portal_type='Document', id='page1')
 

--- a/src/archetypes/referencebrowserwidget/tests/test_product.py
+++ b/src/archetypes/referencebrowserwidget/tests/test_product.py
@@ -283,7 +283,7 @@ class PopupTestCase(PopupBaseTestCase):
         batch = popup.getResult()
         assert isinstance(batch, Batch)
         assert len(batch) == 1
-        assert batch[0].getObject() == self.obj
+        assert batch[0]['item'].getObject() == self.obj
         assert popup.has_queryresults
 
     def test_title_or_id(self):

--- a/src/archetypes/referencebrowserwidget/tests/test_product.py
+++ b/src/archetypes/referencebrowserwidget/tests/test_product.py
@@ -668,15 +668,24 @@ class IntegrationTestCase(FunctionalTestCase):
         else:
             self.assertEqual(len(INSERTLINK.findall(body)), wanted_insertlinks)
 
+        # add a news-item, which is not shown in the popup because its not in allowed_types
         makeContent(self.portal, portal_type='News Item', id='newsitem')
         body = self.getNormalizedPopup()
-
-        self.assertEqual(len(ROWS.findall(body)), wanted_rows + 1)
+        self.assertEqual(len(ROWS.findall(body)), wanted_rows, 'not linkable types should not be shown')
         if HAS_DASH_UUID:
             self.assertEqual(len(INSERTLINK_UUID.findall(body)), wanted_insertlinks)
         else:
             self.assertEqual(len(INSERTLINK.findall(body)), wanted_insertlinks)
-
+            
+        # add a document, this will be addable in the popup
+        makeContent(self.portal, portal_type='Document', id='another-doc')
+        body = self.getNormalizedPopup()
+        self.assertEqual(len(ROWS.findall(body)), wanted_rows + 1)
+        if HAS_DASH_UUID:
+            self.assertEqual(len(INSERTLINK_UUID.findall(body)), wanted_insertlinks + 1)
+        else:
+            self.assertEqual(len(INSERTLINK.findall(body)), wanted_insertlinks + 1)
+            
     def test_bc_navigationroot(self):
         makeContent(self.portal.folder1, portal_type='Document', id='page1')
 

